KubeOne: documentation for Operating System Manager (#1133)
kubermatic/docs: content/kubeone/master/guides/operating-system-manager/_index.en.md

- Each cloud provider sets some sort of limits for the size of `user-data`, machine won't be created in case of non-compliance. For example, at the time of writing this, AWS has set a [hard limit of 16KB](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-add-user-data.html).
- Better support for air-gapped environments is required.

Operating System Manager was created to overcome these limitations.

## Architecture

OSM introduces the following new resources:

### OperatingSystemProfile

Templatized resource that represents the details of each operating system. OSPs are immutable and default OSPs for supported operating systems are provided/installed automatically by kubeOne. End users can create custom OSPs as well to fit their own use-cases.

### OperatingSystemConfig

Immutable resource that contains the **actual configurations** that are going to be used to bootstrap and provision the worker nodes. It is a subset of OperatingSystemProfile, OperatingSystemProfile is a template while OperatingSystemConfig is an instance rendered using OperatingSystemProfile, MachineDeployment and flags provided at OSM level.

> OperatingSystemConfigs have a 1-to-1 relation with the MachineDeployment. A dedicated controller watches the MachineDeployments and generates the OSCs in `kube-system` and secrets in `cloud-init-settings` namespaces in the cluster. Machine Controller then waits for the bootstrapping and provisioning secrets to become available. Once they are ready, it will extract the configurations from those secrets and pass them as `user-data`
> to the machines that it's going to provision.

Reviewer: ```suggestion
OperatingSystemConfigs have a 1-to-1 relation with the MachineDeployment. A dedicated controller watches the MachineDeployments and generates the OSCs in `kube-system` and secrets in `cloud-init-settings` namespaces in the cluster. Machine Controller then waits for the bootstrapping- and provisioning-secrets to become available. Once they are ready, it will extract the configurations from those secrets and pass them as `user-data` to the to-be-provisioned machines.
```

For each MachineDeployment we have two types of configurations, which are stored in secrets:

> 1. **Bootstrap**: Configuration ussed for initially setting up the machine and to fetch the provisioning configuration.

Reviewer: ```suggestion
1. **Bootstrap**: Configuration used for initially setting up the machine and fetching the provisioning configuration.
```
